# Genome Continuity Index
## About
Genome Continuity Index (GCI) is a program for assessing the T2T genome, which has a high resolution (single-base level in theory). After analyzing the alignment files generated by mapping long reads (PacBio HiFi and/or Oxford Nanopore long reads) to the T2T genome, GCI can provide an index to represent the continuity of the genome.

**As for the version 1.0, we can only use one thread, as well as other limits (see [To do](https://github.com/yeeus/GCI#to-do) for more details)**

![7fe814001fd9da8e1ef94fb3f386129](https://github.com/yeeus/genome_assessment_tool/assets/118142448/75b978b6-a29f-4ade-b9c2-51a1c0ff60b0)



## Contents
- [Requirements](https://github.com/yeeus/GCI#requirements)
- [Parameters](https://github.com/yeeus/GCI#parameters)
- [Usage](https://github.com/yeeus/GCI#usage) 
- [Outputs](https://github.com/yeeus/GCI#outputs)
- [Utility](https://github.com/yeeus/GCI#utility)
- [Citation](https://github.com/yeeus/GCI#citation)
- [Help](https://github.com/yeeus/GCI#help)
- [To do](https://github.com/yeeus/GCI#to-do)

### Requirements
For the complete pipeline, there are several necessary softwares:

- [canu](https://github.com/marbl/canu) (for trio-binning)
- [seqkit](https://github.com/shenwei356/seqkit) (for seq processing)
- [minimap2](https://github.com/lh3/minimap2) (for mapping)
- [winnowmap](https://github.com/marbl/Winnowmap) (for mapping)
- [samtools](https://github.com/samtools/samtools) (for sam/bam processing)
- [paftools.js](https://github.com/lh3/minimap2/blob/master/misc/paftools.js) (for converting sam to paf)

As for **GCI**, it requires:
- **python3.x** (written in python3.11.0 and tested in python3.10.8)
- pysam (stable version)
- numpy (stable version)
- [bamsnap](https://github.com/zy041225/bamsnap) (for plotting) ###


### Parameters
```
python GCI.py --help

usage: GCI.py [--hifi  [...]] [--nano  [...]] [-ts INT] [-dp FLOAT] [-d PATH] [-o STR] [-t INT] [-mq INT] [-ip FLOAT] [-op FLOAT]
                 [-cp FLOAT] [-fl INT] [-g] [-f] [-p] [-h] [-v]

A program for assessing the T2T genome

Input/Output:
  --hifi  [ ...]        PacBio HiFi reads alignment files (at least one bam file)
  --nano  [ ...]        Oxford Nanopore long reads alignment files (at least one bam file)
  -ts INT, --threshold INT
                        The threshold of depth in the final bed file [0]
  -dp FLOAT, --dist-percent FLOAT
                        The percentage of the distance between the candidate gap intervals in the whole chromosome (contig) [0.001]
  -d PATH               The directory of output files [.]
  -o STR, --output STR  Prefix of output files [GCI]
  -t INT, --threads INT
                        Number of threads [1]

Filter Options:
  -mq INT, --map-qual INT
                        Minium mapping quality for alignments [30]
  -ip FLOAT, --iden-percent FLOAT
                        Minimum identity (num_match_res/len_aln) of the reads [0.9]
  -op FLOAT, --ovlp-percent FLOAT
                        Minimum overlapping percentage of the reads if inputting more than one alignment files [0.9]
  -cp FLOAT, --clip-percent FLOAT
                        Maximum clipped percentage of the reads [0.1]
  -fl INT, --flank-len INT
                        The flanking length of the clipped bases [10]

Other Options:
  -g, --generate        Generate the depth files
  -f, --force           Force rewriting of existing files
  -p, --plot            Visualize the final result
  -h, --help            Show this help message and exit
  -v, --version         Show program's version number and exit

Examples:
python GCI.py --hifi hifi.bam hifi.paf ... --nano nano.bam nano.paf ...
```

### Usage
1. **(For haplotype-resolved genome)** Prepare parental (specific) reads (if parental sequencing data are available, please skip this step) 
```
# we recommend to use canu for binning
canu -haplotype \
    -p $prefix -d $dictionary \
    genomeSize=3g \
    maxThreads=$threads \
    -haplotypePat $pat \
    -haplotypeMat $mat \
    -pacbio-hifi $hifi \   ## binning ONT reads with -nanopore $ont
    useGrid=false

# because there would be unknown reads which could't be reliably binned, we suggest to combine them with haplotype-specific reads
seqkit shuffle -2 -o ${canu_unknown_shuffle.fa.gz} -j $threads ${canu_unknown.fa.gz}
seqkit split2 ${canu_unknown_shuffle.fa.gz} -p 2 -j $threads
cat ${canu_mat.fa.gz} ${canu_unknown_shuffle.part_001.fa.gz} > ${canu_mat.final.fa.gz}
cat ${canu_pat.fa.gz} ${canu_unknown_shuffle.part_002.fa.gz} > ${canu_pat.final.fa.gz}
```

2. Map HiFi and/or ONT reads to assemblies (using minimap2 and winnowmap)
```
# minimap2 
minimap2 -t $threads -ax map-hifi $mat_asm $mat_hifi > ${mat.minimap2.hifi.sam}   ## mapping ONT reads with -ax map-ont

samtools view -@ $threads -Sb ${mat.minimap2.hifi.sam} | samtools sort -@ $threads -o ${mat.minimap2.hifi.bam}
samtools index ${mat.minimap2.hifi.bam} ## this is necessary!!!
paftools.js sam2paf (-p) ${mat.minimap2.hifi.sam} | sort -k6,6V -k8,8n > ${mat.minimap2.hifi.paf} ## We recommend to use "-p" to filter the supplementary alignments
                                                                                                  ## please sort the paf file
                                                                                                  ## because our program don't automatically sort the file by the targets names!

# winnowmap
meryl count k=15 output $mat_merylDB $mat_asm
meryl print greater-than distinct=0.9998 $mat_merylDB > $mat_repetitive_k15.txt
winnowmap -W $mat_repetitive_k15.txt -ax map-pb $mat_asm $mat_hifi > ${mat.winnowmap.hifi.sam}   ## mapping ONT reads with -ax map-ont

samtools view -@ $threads -Sb ${mat.winnowmap.hifi.sam} | samtools sort -@ $threads -o ${mat.winnowmap.hifi.bam}
samtools index ${mat.minimap2.hifi.bam} ## this is necessary!!!
paftools.js sam2paf (-p) ${mat.winnowmap.hifi.sam} | sort -k6,6V -k8,8n > ${mat.winnowmap.hifi.paf} ## We recommend to use "-p" to filter the supplementary alignments
                                                                                                    ## please sort the paf file
                                                                                                    ## because our program don't automatically sort the file by the targets names!
```

3. Filter the mapping files and get the genome continuity index

We recommend to input only one alignment file per software (minimap2 and winnowmap) using the same set of long reads. **Importantly,** there needs at least one bam file for one type of long reads, which means you'll get errors when providing only paf files.
```
# Before this, make sure you've generated the index file (.bai) for bam files
# we recommend to input one bam and one paf file produced by two softwares (for example, one bam file from minimap2 and one paf file from winnowmap)
python GCI.py --hifi hifi.bam hifi.paf (--nano ont.bam ont.paf) -d mat -o mat -t $threads ... ## as for v.1.0, we can only use one thread
```

### Outputs
#### only providing one type of reads
- ${dictionay}
    - (${prefix}.depth) ## the whole-genome depth file (only generated with the parameter **-g**)
    - ${prefix}.${threshold}.depth.bed ## the merged depth file in bed format
    - ${prefix}.gci ## an index file containing the reads type, expected N50, observed N50, expected number of contigs, observed number of contigs, genome continuity index
  
#### providing two types
- ${dictionay}
  - (${prefix}_hifi.depth) ## the whole-genome depth file generated by the hifi alignment file
  - (${prefix}_nano.depth) ## the whole-genome depth file generated by the ont alignment file
  - (${prefix}_two_type.depth) ## the whole-genome depth file generated by the two types of reads alignment file
  - ${prefix}_hifi.${threshold}.depth.bed ## the merged depth file in bed format generated by the hifi alignment file
  - ${prefix}_nano.${threshold}.depth.bed ## the merged depth file in bed format generated by the ont alignment file
  - ${prefix}_two_type.${threshold}.depth.bed ## the merged depth file in bed format generated by the two types of reads alignment file
  - ${prefix}.gci

### Utility
- filter_bam.py
  - Parameters
    ```
    usage: filter_bam.py [-d PATH] [-o [STR ...]] [-t INT] [-mq INT] [-ip FLOAT] [-op FLOAT] [-cp FLOAT] [-p] [-r STR] [-R FILE] [-f] [-h]
                     ALIGNMENT-FILE [ALIGNMENT-FILE ...]

    This is the core part of the main program GCI.py with additional functions
    You can provide some bam and paf files (like GCI.py) and set the filtered parameters
    and then you will get the filtered bam file which can be used for subsequent analysis
    
    Input/Output:
      ALIGNMENT-FILE        Long reads alignment files (at least one bam file)
      -d PATH               The directory of output files [.]
      -o [STR ...], --output [STR ...]
                            Prefix of output files; one prefix corresponds to one bam file in order and if provide the parameter "-p", the last
                            one was used as the prefix for bamsnap outputs [[$input.filter] for filtered bam files and [GCI] for bamsnap
                            outputs]
      -t INT, --threads INT
                            Number of threads [1]
    
    Filter Options:
      -mq INT, --map-qual INT
                            Minium mapping quality for alignments [30]
      -ip FLOAT, --iden-percent FLOAT
                            Minimum identity (num_match_res/len_aln) of the reads [0.9]
      -op FLOAT, --ovlp-percent FLOAT
                            Minimum overlapping percentage of the reads if inputting more than one alignment files [0.9]
      -cp FLOAT, --clip-percent FLOAT
                            Maximum clipped percentage of the reads [0.1]
    
    Plot Options:
      -p, --plot            Visualize the filtered bam files
      -r STR, --region STR  The region to plot in chr:pos or chr:start-end format
      -R FILE, --regions-file FILE
                            Bed file contains the regions to plot
    
    Other Options:
      -f, --force           Force rewriting of existing files
      -h, --help            Show this help message and exit
    
    Examples:
    python filter_bam.py bam1 paf1 ...
    ```
  - Usage

    This script is used to get the filtered bam files (and get the final plots).
    - Example1

      After getting the ${prefix}.${threshold}.depth.bed file, we'd like to get the detailed filtering information. So, we can just extract the alignments in the bed file:
      ```
      samtools view -@ $threads -Sb -L ${prefix}.${threshold}.depth.bed ${hifi.bam} > test.bam
      samtools index test.bam
      ```
      Then we can input the bam file with other paf file(s):
      ```
      python filter_bam.py test.bam test.paf -d test -o test ## if no prefix provided, the output file would be `test.filter.bam`
      ```
      Finally we get the filtered bam file `test.bam`.

      After `samtools sort && index`, we can visualize the sorted file in [IGV](https://github.com/igvteam/igv):
      > ![igv_test](https://github.com/yeeus/GCI/images/igv_test.png)

    - Example2

      We can immediately visualize the alignments after getting the filtered file.
      
### Citation


### Help
If you get any problems, please raise an [issue](https://github.com/yeeus/GCI/issues) first.
For another helps, please contact quanyu_chen@outlook.com.


### To do
- visualization
- contribution for N50 per chromosome
- multiple threads
- simulated
